qi-lib/flow/core/runtime.rkt

```diff
@@ -171,25 +172,36 @@
 
 (define for-all andmap)
 
+(define (singleton? seq)
+  ;; cheap check to see if a list is of length 1,
+  ;; instead of traversing to compute the length
+  (and (not (empty? seq))
+       (empty? (rest seq))))
+
 (define (~zip-with op seqs truncate)
-  (if (empty? seqs)
-      null
-      (if (exists empty? seqs)
-       (if (for-all empty? seqs)
-           null
-           (if truncate
-               null
-               (apply raise-arity-error
-                      'zip-with
-                      0
-                      (first (filter (negate empty?) seqs)))))
-       (let ([vs (map first seqs)])
-         (append (values->list (apply op vs))
-                 (~zip-with op (map rest seqs) truncate))))))
+  (if (exists empty? seqs)
```

Comment:
Probably a slight optimization over doing the null check ourselves, seems reasonable. Notably the empty check also appears in the caller `zip-with`.

Reply:
The caller null check checks whether `seqs` is empty, whereas this one checks if any of the members of `seqs` is empty. We also have the caller _of_ the caller do what feels like a redundant `singleton?` check before re-packing the arguments into a list (in the id case of `sep`). I seemed to conclude that was necessary because if we do it within `zip-with`, we already have lost the "fast path" option for the common case. I jumbled things around for a while before settling on this implementation somewhat empirically, but I wouldn't be surprised if it could be improved.